### PR TITLE
Add unit tests for StorageKey classes.

### PR DESCRIPTION
**Pull request by jasonwyatt**
Also adds parsing capability for ReferenceModeStorageKey - something that was done by Shane in #4000.

Ref #4037

### CODE REVIEW

File: src_kt/java/arcs/storage/BUILD [hunk 1]
```diff
@@ -36,6 +37,7 @@ kt_jvm_and_js_library(
     name = "storage_key",
     srcs = [
         "StorageKey.kt",
+        "StorageKeyParser.kt",
     ],
```
Comment (csilvestrini) [inline src_kt/java/arcs/storage/BUILD]: Optional: You could define a `STORAGE_KEY_SRCS` const for these, and then you wouldn't need to be so careful to duplicated the same list of srcs here and in `excludes` above.

File: src_kt/java/arcs/storage/driver/RamDisk.kt [hunk 2]
```diff
@@ -12,20 +12,42 @@
...
 
 /** Storage key for a piece of data managed by the ramdisk driver. */
 data class RamDiskStorageKey(private val unique: String) : StorageKey(RAMDISK_DRIVER_PROTOCOL) {
-  override fun toKeyString(): String = unique
+    override fun toKeyString(): String = unique
+
+    override fun childKeyWithComponent(component: String): StorageKey =
+        RamDiskStorageKey("$unique/$component")
+
+    override fun toString(): String = super.toString()
```
Comment (csilvestrini) [inline src_kt/java/arcs/storage/driver/RamDisk.kt]: Is this needed? Isn't that the default behaviour?

Comment (jasonwyatt) [reply src_kt/java/arcs/storage/driver/RamDisk.kt]: Unfortunately it is needed. I think it has something to do with data classes implementing `toString` for you, which - while normally helpful - isn't really what we want here.

File: src_kt/java/arcs/storage/BUILD (hunk 1 above)
Comment (jasonwyatt) [reply src_kt/java/arcs/storage/BUILD]: Thanks, nice tip!

**Pull request merged**